Fix deadlines handler

thed636 opened the pull request:
## PROBLEM

Both callbacks (normal flow and cancel timer) appear in the io_context queue so both are called. No such proper handling for the situation.

## FIX

Add a shared state to prevent both callbacks execution. The backward atomic counter is used to indicate the calling sequence.
Error code for timed out operation now is `boost::asio::error::timed_out`.

## Add-on

Make `get_database`, `get_host`, `get_port`, `get_user`, `get_password` compatible with null handler

In `libpq` corresponding functions return `nullptr` for `nullptr` connection handle, `std::string_view` does not handle `nullptr` for `const char*` argument, so it leads to UB. This commit makes OZO functions to behave like the underlying `libpq` functions - return empty `std::string_view` for `nullptr` native connection handle.

thed636 committed:
Fix deadlines handler

PROBLEM

Both callbacks (normal flow and cancel timer) appear in io_context queue
so both are called. No such proper handling for the situation.

FIX

Add shared state to prevent both callbacks execution.
Error code for timed out operation now is `boost::asio::error::timed_out`.

serenityFX commented on include/ozo/detail/deadline.h:
```diff
@@ -33,32 +38,44 @@ class deadline_handler {
         void operator() (error_code ec) {
-            if (ec != asio::error::operation_aborted) {
-                asio::dispatch(ozo::detail::bind(std::move(target_), std::move(ec)));
+            if (--(ctx_->state_) && ec != asio::error::operation_aborted) {
```
if ec = asio::error::operation_aborted call the wrong handler because state was decremented

thed636 replied:
Yep, thanks.
**Done**

Commit:
Make gat_database, get_host, get_port, get_user, get_password compatible with null handler

In libpq corresponding functions return nullptr for nullptr connection handle,
std::string_view does not nullptr for const char* argument, so it leads to UB.
This commit makes OZO functions to behave as the underlying libpq functions -
return empty std::string_view for nullptr native connection handle.
